File: 2024/py/day01.py
import os


with open(os.path.join(os.path.abspath(os.path.dirname(__file__)), "../input/day01.txt"), "r") as f:
    l = []
    r = []
    for line in f.readlines():
        split_line = line.split()
        l.append(int(split_line[0]))
        r.append(int(split_line[1]))

l.sort()
r.sort()

# p1

sum = sum(abs(a - b) for (a, b) in zip(l, r))
print(sum)

# p2

s = {}
for id in l:
    if id in s:
        continue

    count = 0
    for rep in r:
        if rep == id:
            count += 1
    s[id] = count

sum = 0
for (id, count) in s.items():
    sum += (id * count)

print(sum)

File: 2024/bin/day01.ml
open StdLabels

let input = Aoc.read_input_lines 1

let split line = String.split_on_char ~sep:' ' line |> List.filter ~f:(fun s -> s <> "")

let rec split_list list l r =
  match list with
  | [] -> (l, r)
  | hd :: rest -> (
      match split hd with
      | [ a; b ] -> split_list rest (int_of_string a :: l) (int_of_string b :: r)
      | _ -> Aoc.raise_input_failure ())

let l, r = split_list input [] []

module IntList = struct
  let sum = List.fold_left ~init:0
  let sum2 = List.fold_left2 ~init:0
  let sort = List.sort ~cmp:compare
  let count ~value = sum ~f:(fun acc i -> if i == value then acc + 1 else acc)
end

(* part 1 *)
let () =
  let sum = IntList.sum2 ~f:(fun acc a b -> acc + abs (a - b)) (IntList.sort l) (IntList.sort r) in
  print_endline @@ string_of_int sum

(* part 2 *)
let () =
  let sum = IntList.sum ~f:(fun acc id -> acc + (id * IntList.count ~value:id r)) l in
  print_endline @@ string_of_int sum
